- Handle change in config file format

rjschwei opened the pull request:
+ Config file to be processed changed to list of dicts from dict
    to handle same repo for multiple frameworks

rjschwei committed:
- Handle change in config file format
  + Config file to be processed changed to list of dicts from dict
    to handle same repo for multiple frameworks

ikapelyukhin commented:
Hmm, where has the config format changed?

rjschwei commented:
> Hmm, where has the config format changed?

In https://github.com/SUSE/pubcloud/pull/418

ikapelyukhin commented:
> > Hmm, where has the config format changed?
> 
> In [SUSE/pubcloud#418](https://github.com/SUSE/pubcloud/pull/418)

But it doesn't seem like `rmt_repos_config.json` format has changed there?

rjschwei commented:
> > > Hmm, where has the config format changed?
> > 
> > 
> > In [SUSE/pubcloud#418](https://github.com/SUSE/pubcloud/pull/418)
> 
> But it doesn't seem like `rmt_repos_config.json` format has changed there?

You are correct, forgot the PR that changes the format. 4 branches across 3 repos, one PR got lost :(

https://github.com/SUSE-Enceladus/update-infra-utils/pull/5

changes the format.

ikapelyukhin commented on check_rmt_repos:
```diff
@@ -108,24 +108,27 @@ extra_enabled_repos = []
+                missing_repo_dirs.append(full_repo_path)
+    
+            if enabled_repos.get(id):
+                del enabled_repos[id]
```
This is broken now because `config_repos` is an array now instead of a hash, a repo can be repeated more than once -- it will be deleted on the first encounter of the repo and report that the repo is missing next time it is encountered in the list.

ikapelyukhin commented on check_rmt_repos:
```diff
@@ -108,24 +108,27 @@ extra_enabled_repos = []
-    extra_mirrored_repos = list(mirrored_repos.keys())
+    for config_repo in config_repos:
+        for id, repo_info in config_repo.items():
+            if not enabled_repos.get(id):
```
It doesn't seem to be comparing host system framework vs. repo framework that's in the config anywhere -- it will complain about repos missing that belong to other frameworks.

rjschwei committed:
- Repo checker
  + Fix logic to only consider non framework specific repos and those
    repos configured for this framework. Requires new "--cloud" command line
    argument.
  + Additional change to handle the config file format change from dict
    to list

ikapelyukhin commented:
@rjschwei let's use the description column (`repo_detail[2]`) -- repo arch is included in the description.
https://github.com/SUSE-Enceladus/monitoring/blob/f5032cfcf076cfab1c06d2082cb90dac5b35e9d6/check_rmt_repos#L50

ikapelyukhin commented:
It probably would be a good idea to make the output usable for cleaning up instances with config issues:
1. Output the list of space separated repo IDs that need to be disabled, so that they could be easily fed into `rmt-cli repos disable`.
2. Output a list of extra mirrored repo locations one per line, so that they can be removed easily.

rjschwei committed:
- Improve the data output in case of inconsistency detections
  + Previously we just dumped lists of data. This made things hard to parse
    by the person reviewing the output and thus it was almost impossible to
    detect if the issues were real or false positives. Also dump a list
    of repo ids to make it easy to use the information for the rmt-cli
    command.

rjschwei commented:
@ikapelyukhin suggesting implemented, still some other issues per e-mail

ikapelyukhin commented on check_rmt_repos:
```diff
@@ -87,6 +87,14 @@ argparse.add_argument(
+help_msg = 'A designator that will compare the repository configuration '
+help_msg += '"framework" setting with the provided value'
+argparse.add_argument(
+    '--cloud',
```
Maybe we should detect it automatically, like we do in repo enablement script?

rjschwei replied:
@ikapelyukhin The repo script is internal and therefore can be targeted to our specific use cases. However, we expect CSPs that run their own infrastructure to be able to use this public tool and therefore hard coding our use cases would work against the general purpose nature of this code.

ikapelyukhin replied:
Yes, but:

* We have automatic detection in `enable-rmt-repos` script, which is also supposed to be usable by 3rd parties.
* This is a mandatory setting now -- it doesn't make any sense for 3rd party CSPs, as they are not operating it on multiple different frameworks. They don't have to explicitly set it in the product config at all, most likely they'll have to supply some random value for this script.
* This makes our monitoring configs CSP-specific.

Can we instead do auto-detection, and if no framework is detected -- ignore all framework-specific repos? That'd work for us and for 3rd party CSPs.

ikapelyukhin replied:
Repo enablement script w/ autodetection isn't internal: https://github.com/SUSE-Enceladus/update-infra-utils/blob/master/update-server-repo-setup/enable-rmt-repos#L87-L94

rjschwei replied:
Fair point and we have to fix this. However, a 3rd party CSP may still have multiple environments where different repositories get enabled. I'll rework the handling of the framework in both places. Thanks

ikapelyukhin replied:
OK, requiring it to be set explicitly in the repo enablement script would also work. :slightly_smiling_face:

ikapelyukhin reviewed (approved):
But other than the framework discussion -- LGTM. :+1:

rjschwei committed:
- Do not make the new "--cloud" argument required
  + It is perfectly reasonable to have a repo configuration that has no
    cloud specific components
- Bump he version to the actual release date
